ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -673,8 +672,7 @@ Operator genOPTree(ASTNode ast, PlannerContext plannerCtx) throws SemanticExcept
           }
           this.ctx.setCboInfo(cboMsg);
 
-          // Determine if we should re-throw the exception OR if we try to mark plan as reAnalyzeAST to retry
-          // planning as non-CBO.
+          // Determine if we should re-throw the exception OR if we try to mark the query to retry as non-CBO.
```

Comment:
I am wondering if it would be better to just throw here and let the reexecution plugin deal with what do to afterwards.

Reply:
That was my first implementation, but that messed up with the current query outputs (qfiles).
I have tried to throw the ReCompileException (CBOException at that time), and then decide based on the exception. The problem arisen when I had to extract and rethrow the cause of the CBOException if the exception was fatal to provide the same outputs. I have chickened out from that path to keep the existing logic in place.

Reply:
Interesting, I was about to comment that `ReCompileException` should better be `CBOException` :) It really feels that this is a better approach. 

If we review this part of the code independently of what happens elsewhere it seems reasonable that we are catching an exception and wrapping it around another more "high-level" abstraction. This is usually a good coding pattern.

Apart from that, I don't know if it is crucial to retain exactly the same outputs in tests. For instance, if the changes you mention are due to the following part of the code:
```
            // Wrap all other errors (Should only hit in tests)
            throw new SemanticException(e);
```
then it makes believe even more that we shouldn't care.

Reply:
@pvary have explored this in the first patch - and we were getting into changing error message/exception types/etc; because of the tighter co-location of the logic and the recompilation stuff

I suggest to separate those into a new separate changeset and get this in separately